mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoGetEntityMethodGenerator.java

```diff
@@ -0,0 +1,145 @@
+import com.datastax.oss.driver.internal.mapper.processor.MethodGenerator;
+import com.datastax.oss.driver.internal.mapper.processor.ProcessorContext;
+import com.datastax.oss.driver.internal.mapper.processor.SkipGenerationException;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.MethodSpec;
+import java.util.List;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+
+public class DaoGetEntityMethodGenerator implements MethodGenerator {
+
+  private final ExecutableElement methodElement;
+  private final DaoImplementationGenerator daoImplementationGenerator;
+  private final String targetParameterName;
+  private final TypeElement entityElement;
+  private boolean pagingReturnType;
+
+  public DaoGetEntityMethodGenerator(
+      ExecutableElement methodElement,
+      DaoImplementationGenerator daoImplementationGenerator,
+      ProcessorContext context) {
+    this.methodElement = methodElement;
+    this.daoImplementationGenerator = daoImplementationGenerator;
+    // Methods should only have one parameter
+    if (methodElement.getParameters().size() != 1) {
+      context
+          .getMessager()
+          .error(
+              methodElement, "%s method must have one parameter", GetEntity.class.getSimpleName());
+      throw new SkipGenerationException();
+    }
+    // Parameter name as a string
+    String tmpParameterString = null;
+    // The return type as an element
+    TypeElement returnTypeElement = null;
+    VariableElement parameterElement = methodElement.getParameters().get(0);
+    TypeMirror parameterType = parameterElement.asType();
+    // Check the parameter type make sure it matches an expected type
+    if (context.getClassUtils().implementsGettableByName(parameterType)
+        || context.getClassUtils().isSame(parameterType, ResultSet.class)
+        || context.getClassUtils().isSame(parameterType, AsyncResultSet.class)) {
+      tmpParameterString = parameterElement.getSimpleName().toString();
+    }
+    // Check return type. Make sure it matches the expected type
+    TypeMirror returnType = methodElement.getReturnType();
+    if (returnType.getKind() == TypeKind.DECLARED) {
+      Element element = ((DeclaredType) returnType).asElement();
+      // Simple case return type is an entity type
+      if (element.getKind() == ElementKind.CLASS) {
+        if (element.getAnnotation(Entity.class) != null) {
+          returnTypeElement = ((TypeElement) ((DeclaredType) returnType).asElement());
+        }
+        // Return type is a an iterable type, check nested generic as well
+      } else if (element.getKind() == ElementKind.INTERFACE) {
```

Comment:
That's done below, but I agree that it would be a bit more clear to do all the checks on the main type first, and only then inspect the type parameters. I'll change it.